// @aws-sdk/client-resource-groups-tagging-api/dist-cjs/index.js
'use strict';

var middlewareHostHeader = require('@aws-sdk/middleware-host-header');
var middlewareLogger = require('@aws-sdk/middleware-logger');
var middlewareRecursionDetection = require('@aws-sdk/middleware-recursion-detection');
var middlewareUserAgent = require('@aws-sdk/middleware-user-agent');
var configResolver = require('@smithy/config-resolver');
var core = require('@smithy/core');
var schema = require('@smithy/core/schema');
var middlewareContentLength = require('@smithy/middleware-content-length');
var middlewareEndpoint = require('@smithy/middleware-endpoint');
var middlewareRetry = require('@smithy/middleware-retry');
var smithyClient = require('@smithy/smithy-client');
var httpAuthSchemeProvider = require('./auth/httpAuthSchemeProvider');
var runtimeConfig = require('./runtimeConfig');
var regionConfigResolver = require('@aws-sdk/region-config-resolver');
var protocolHttp = require('@smithy/protocol-http');

const resolveClientEndpointParameters = (options) => {
    return Object.assign(options, {
        useDualstackEndpoint: options.useDualstackEndpoint ?? false,
        useFipsEndpoint: options.useFipsEndpoint ?? false,
        defaultSigningName: "tagging",
    });
};
const commonParams = {
    UseFIPS: { type: "builtInParams", name: "useFipsEndpoint" },
    Endpoint: { type: "builtInParams", name: "endpoint" },
    Region: { type: "builtInParams", name: "region" },
    UseDualStack: { type: "builtInParams", name: "useDualstackEndpoint" },
};

const getHttpAuthExtensionConfiguration = (runtimeConfig) => {
    const _httpAuthSchemes = runtimeConfig.httpAuthSchemes;
    let _httpAuthSchemeProvider = runtimeConfig.httpAuthSchemeProvider;
    let _credentials = runtimeConfig.credentials;
    return {
        setHttpAuthScheme(httpAuthScheme) {
            const index = _httpAuthSchemes.findIndex((scheme) => scheme.schemeId === httpAuthScheme.schemeId);
            if (index === -1) {
                _httpAuthSchemes.push(httpAuthScheme);
            }
            else {
                _httpAuthSchemes.splice(index, 1, httpAuthScheme);
            }
        },
        httpAuthSchemes() {
            return _httpAuthSchemes;
        },
        setHttpAuthSchemeProvider(httpAuthSchemeProvider) {
            _httpAuthSchemeProvider = httpAuthSchemeProvider;
        },
        httpAuthSchemeProvider() {
            return _httpAuthSchemeProvider;
        },
        setCredentials(credentials) {
            _credentials = credentials;
        },
        credentials() {
            return _credentials;
        },
    };
};
const resolveHttpAuthRuntimeConfig = (config) => {
    return {
        httpAuthSchemes: config.httpAuthSchemes(),
        httpAuthSchemeProvider: config.httpAuthSchemeProvider(),
        credentials: config.credentials(),
    };
};

const resolveRuntimeExtensions = (runtimeConfig, extensions) => {
    const extensionConfiguration = Object.assign(regionConfigResolver.getAwsRegionExtensionConfiguration(runtimeConfig), smithyClient.getDefaultExtensionConfiguration(runtimeConfig), protocolHttp.getHttpHandlerExtensionConfiguration(runtimeConfig), getHttpAuthExtensionConfiguration(runtimeConfig));
    extensions.forEach((extension) => extension.configure(extensionConfiguration));
    return Object.assign(runtimeConfig, regionConfigResolver.resolveAwsRegionExtensionConfiguration(extensionConfiguration), smithyClient.resolveDefaultRuntimeConfig(extensionConfiguration), protocolHttp.resolveHttpHandlerRuntimeConfig(extensionConfiguration), resolveHttpAuthRuntimeConfig(extensionConfiguration));
};

class ResourceGroupsTaggingAPIClient extends smithyClient.Client {
    config;
    constructor(...[configuration]) {
        const _config_0 = runtimeConfig.getRuntimeConfig(configuration || {});
        super(_config_0);
        this.initConfig = _config_0;
        const _config_1 = resolveClientEndpointParameters(_config_0);
        const _config_2 = middlewareUserAgent.resolveUserAgentConfig(_config_1);
        const _config_3 = middlewareRetry.resolveRetryConfig(_config_2);
        const _config_4 = configResolver.resolveRegionConfig(_config_3);
        const _config_5 = middlewareHostHeader.resolveHostHeaderConfig(_config_4);
        const _config_6 = middlewareEndpoint.resolveEndpointConfig(_config_5);
        const _config_7 = httpAuthSchemeProvider.resolveHttpAuthSchemeConfig(_config_6);
        const _config_8 = resolveRuntimeExtensions(_config_7, configuration?.extensions || []);
        this.config = _config_8;
        this.middlewareStack.use(schema.getSchemaSerdePlugin(this.config));
        this.middlewareStack.use(middlewareUserAgent.getUserAgentPlugin(this.config));
        this.middlewareStack.use(middlewareRetry.getRetryPlugin(this.config));
        this.middlewareStack.use(middlewareContentLength.getContentLengthPlugin(this.config));
        this.middlewareStack.use(middlewareHostHeader.getHostHeaderPlugin(this.config));
        this.middlewareStack.use(middlewareLogger.getLoggerPlugin(this.config));
        this.middlewareStack.use(middlewareRecursionDetection.getRecursionDetectionPlugin(this.config));
        this.middlewareStack.use(core.getHttpAuthSchemeEndpointRuleSetPlugin(this.config, {
            httpAuthSchemeParametersProvider: httpAuthSchemeProvider.defaultResourceGroupsTaggingAPIHttpAuthSchemeParametersProvider,
            identityProviderConfigProvider: async (config) => new core.DefaultIdentityProviderConfig({
                "aws.auth#sigv4": config.credentials,
            }),
        }));
        this.middlewareStack.use(core.getHttpSigningPlugin(this.config));
    }
    destroy() {
        super.destroy();
    }
}

class ResourceGroupsTaggingAPIServiceException extends smithyClient.ServiceException {
    constructor(options) {
        super(options);
        Object.setPrototypeOf(this, ResourceGroupsTaggingAPIServiceException.prototype);
    }
}

class ConcurrentModificationException extends ResourceGroupsTaggingAPIServiceException {
    name = "ConcurrentModificationException";
    $fault = "client";
    Message;
    constructor(opts) {
        super({
            name: "ConcurrentModificationException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, ConcurrentModificationException.prototype);
        this.Message = opts.Message;
    }
}
class ConstraintViolationException extends ResourceGroupsTaggingAPIServiceException {
    name = "ConstraintViolationException";
    $fault = "client";
    Message;
    constructor(opts) {
        super({
            name: "ConstraintViolationException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, ConstraintViolationException.prototype);
        this.Message = opts.Message;
    }
}
class InternalServiceException extends ResourceGroupsTaggingAPIServiceException {
    name = "InternalServiceException";
    $fault = "server";
    Message;
    constructor(opts) {
        super({
            name: "InternalServiceException",
            $fault: "server",
            ...opts,
        });
        Object.setPrototypeOf(this, InternalServiceException.prototype);
        this.Message = opts.Message;
    }
}
class InvalidParameterException extends ResourceGroupsTaggingAPIServiceException {
    name = "InvalidParameterException";
    $fault = "client";
    Message;
    constructor(opts) {
        super({
            name: "InvalidParameterException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, InvalidParameterException.prototype);
        this.Message = opts.Message;
    }
}
class ThrottledException extends ResourceGroupsTaggingAPIServiceException {
    name = "ThrottledException";
    $fault = "client";
    Message;
    constructor(opts) {
        super({
            name: "ThrottledException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, ThrottledException.prototype);
        this.Message = opts.Message;
    }
}
class PaginationTokenExpiredException extends ResourceGroupsTaggingAPIServiceException {
    name = "PaginationTokenExpiredException";
    $fault = "client";
    Message;
    constructor(opts) {
        super({
            name: "PaginationTokenExpiredException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, PaginationTokenExpiredException.prototype);
        this.Message = opts.Message;
    }
}

const _CD = "ComplianceDetails";
const _CFRT = "CloudFormationResourceTypes";
const _CME = "ConcurrentModificationException";
const _CS = "ComplianceStatus";
const _CVE = "ConstraintViolationException";
const _DRC = "DescribeReportCreation";
const _DRCI = "DescribeReportCreationInput";
const _DRCO = "DescribeReportCreationOutput";
const _EC = "ErrorCode";
const _ECR = "ExcludeCompliantResources";
const _EM = "ErrorMessage";
const _FI = "FailureInfo";
const _FRM = "FailedResourcesMap";
const _GB = "GroupBy";
const _GCS = "GetComplianceSummary";
const _GCSI = "GetComplianceSummaryInput";
const _GCSO = "GetComplianceSummaryOutput";
const _GR = "GetResources";
const _GRI = "GetResourcesInput";
const _GRO = "GetResourcesOutput";
const _GTK = "GetTagKeys";
const _GTKI = "GetTagKeysInput";
const _GTKO = "GetTagKeysOutput";
const _GTV = "GetTagValues";
const _GTVI = "GetTagValuesInput";
const _GTVO = "GetTagValuesOutput";
const _ICD = "IncludeComplianceDetails";
const _IPE = "InvalidParameterException";
const _ISE = "InternalServiceException";
const _K = "Key";
const _KWNV = "KeysWithNoncompliantValues";
const _LRT = "ListRequiredTags";
const _LRTI = "ListRequiredTagsInput";
const _LRTO = "ListRequiredTagsOutput";
const _LU = "LastUpdated";
const _M = "Message";
const _MR = "MaxResults";
const _NCR = "NonCompliantResources";
const _NK = "NoncompliantKeys";
const _NT = "NextToken";
const _PT = "PaginationToken";
const _PTEE = "PaginationTokenExpiredException";
const _R = "Region";
const _RARN = "ResourceARN";
const _RARNL = "ResourceARNList";
const _RF = "RegionFilters";
const _RPP = "ResourcesPerPage";
const _RT = "RequiredTags";
const _RTF = "ResourceTypeFilters";
const _RTFLRT = "RequiredTagsForListRequiredTags";
const _RTK = "ReportingTagKeys";
const _RTM = "ResourceTagMapping";
const _RTML = "ResourceTagMappingList";
const _RTe = "RequiredTag";
const _RTes = "ResourceType";
const _S = "Status";
const _SB = "S3Bucket";
const _SC = "StatusCode";
const _SD = "StartDate";
const _SL = "S3Location";
const _SLu = "SummaryList";
const _SRC = "StartReportCreation";
const _SRCI = "StartReportCreationInput";
const _SRCO = "StartReportCreationOutput";
const _Su = "Summary";
const _T = "Tags";
const _TE = "ThrottledException";
const _TF = "TagFilters";
const _TFL = "TagFilterList";
const _TFa = "TagFilter";
const _TI = "TargetId";
const _TIF = "TargetIdFilters";
const _TIT = "TargetIdType";
const _TK = "TagKeys";
const _TKF = "TagKeyFilters";
const _TL = "TagList";
const _TPP = "TagsPerPage";
const _TR = "TagResources";
const _TRI = "TagResourcesInput";
const _TRO = "TagResourcesOutput";
const _TV = "TagValues";
const _Ta = "Tag";
const _UR = "UntagResources";
const _URI = "UntagResourcesInput";
const _URO = "UntagResourcesOutput";
const _V = "Value";
const _Va = "Values";
const _c = "client";
const _e = "error";
const _h = "http";
const _s = "server";
const _sm = "smithy.ts.sdk.synthetic.com.amazonaws.resourcegroupstaggingapi";
const n0 = "com.amazonaws.resourcegroupstaggingapi";
var ComplianceDetails$ = [3, n0, _CD,
    0,
    [_NK, _KWNV, _CS],
    [64 | 0, 64 | 0, 2]
];
var ConcurrentModificationException$ = [-3, n0, _CME,
    { [_e]: _c },
    [_M],
    [0]
];
schema.TypeRegistry.for(n0).registerError(ConcurrentModificationException$, ConcurrentModificationException);
var ConstraintViolationException$ = [-3, n0, _CVE,
    { [_e]: _c },
    [_M],
    [0]
];
schema.TypeRegistry.for(n0).registerError(ConstraintViolationException$, ConstraintViolationException);
var DescribeReportCreationInput$ = [3, n0, _DRCI,
    0,
    [],
    []
];
var DescribeReportCreationOutput$ = [3, n0, _DRCO,
    0,
    [_S, _SL, _SD, _EM],
    [0, 0, 0, 0]
];
var FailureInfo$ = [3, n0, _FI,
    0,
    [_SC, _EC, _EM],
    [1, 0, 0]
];
var GetComplianceSummaryInput$ = [3, n0, _GCSI,
    0,
    [_TIF, _RF, _RTF, _TKF, _GB, _MR, _PT],
    [64 | 0, 64 | 0, 64 | 0, 64 | 0, 64 | 0, 1, 0]
];
var GetComplianceSummaryOutput$ = [3, n0, _GCSO,
    0,
    [_SLu, _PT],
    [() => SummaryList, 0]
];
var GetResourcesInput$ = [3, n0, _GRI,
    0,
    [_PT, _TF, _RPP, _TPP, _RTF, _ICD, _ECR, _RARNL],
    [0, () => TagFilterList, 1, 1, 64 | 0, 2, 2, 64 | 0]
];
var GetResourcesOutput$ = [3, n0, _GRO,
    0,
    [_PT, _RTML],
    [0, () => ResourceTagMappingList]
];
var GetTagKeysInput$ = [3, n0, _GTKI,
    0,
    [_PT],
    [0]
];
var GetTagKeysOutput$ = [3, n0, _GTKO,
    0,
    [_PT, _TK],
    [0, 64 | 0]
];
var GetTagValuesInput$ = [3, n0, _GTVI,
    0,
    [_PT, _K],
    [0, 0]
];
var GetTagValuesOutput$ = [3, n0, _GTVO,
    0,
    [_PT, _TV],
    [0, 64 | 0]
];
var InternalServiceException$ = [-3, n0, _ISE,
    { [_e]: _s },
    [_M],
    [0]
];
schema.TypeRegistry.for(n0).registerError(InternalServiceException$, InternalServiceException);
var InvalidParameterException$ = [-3, n0, _IPE,
    { [_e]: _c },
    [_M],
    [0]
];
schema.TypeRegistry.for(n0).registerError(InvalidParameterException$, InvalidParameterException);
var ListRequiredTagsInput$ = [3, n0, _LRTI,
    0,
    [_NT, _MR],
    [0, 1]
];
var ListRequiredTagsOutput$ = [3, n0, _LRTO,
    0,
    [_RT, _NT],
    [() => RequiredTagsForListRequiredTags, 0]
];
var PaginationTokenExpiredException$ = [-3, n0, _PTEE,
    { [_e]: _c },
    [_M],
    [0]
];
schema.TypeRegistry.for(n0).registerError(PaginationTokenExpiredException$, PaginationTokenExpiredException);
var RequiredTag$ = [3, n0, _RTe,
    0,
    [_RTes, _CFRT, _RTK],
    [0, 64 | 0, 64 | 0]
];
var ResourceTagMapping$ = [3, n0, _RTM,
    0,
    [_RARN, _T, _CD],
    [0, () => TagList, () => ComplianceDetails$]
];
var StartReportCreationInput$ = [3, n0, _SRCI,
    0,
    [_SB],
    [0]
];
var StartReportCreationOutput$ = [3, n0, _SRCO,
    0,
    [],
    []
];
var Summary$ = [3, n0, _Su,
    0,
    [_LU, _TI, _TIT, _R, _RTes, _NCR],
    [0, 0, 0, 0, 0, 1]
];
var Tag$ = [3, n0, _Ta,
    0,
    [_K, _V],
    [0, 0]
];
var TagFilter$ = [3, n0, _TFa,
    0,
    [_K, _Va],
    [0, 64 | 0]
];
var TagResourcesInput$ = [3, n0, _TRI,
    0,
    [_RARNL, _T],
    [64 | 0, 128 | 0]
];
var TagResourcesOutput$ = [3, n0, _TRO,
    0,
    [_FRM],
    [() => FailedResourcesMap]
];
var ThrottledException$ = [-3, n0, _TE,
    { [_e]: _c },
    [_M],
    [0]
];
schema.TypeRegistry.for(n0).registerError(ThrottledException$, ThrottledException);
var UntagResourcesInput$ = [3, n0, _URI,
    0,
    [_RARNL, _TK],
    [64 | 0, 64 | 0]
];
var UntagResourcesOutput$ = [3, n0, _URO,
    0,
    [_FRM],
    [() => FailedResourcesMap]
];
var ResourceGroupsTaggingAPIServiceException$ = [-3, _sm, "ResourceGroupsTaggingAPIServiceException", 0, [], []];
schema.TypeRegistry.for(_sm).registerError(ResourceGroupsTaggingAPIServiceException$, ResourceGroupsTaggingAPIServiceException);
var RequiredTagsForListRequiredTags = [1, n0, _RTFLRT,
    0, () => RequiredTag$
];
var ResourceTagMappingList = [1, n0, _RTML,
    0, () => ResourceTagMapping$
];
var SummaryList = [1, n0, _SLu,
    0, () => Summary$
];
var TagFilterList = [1, n0, _TFL,
    0, () => TagFilter$
];
var TagList = [1, n0, _TL,
    0, () => Tag$
];
var FailedResourcesMap = [2, n0, _FRM,
    0, 0, () => FailureInfo$
];
var DescribeReportCreation$ = [9, n0, _DRC,
    { [_h]: ["POST", "/DescribeReportCreation", 200] }, () => DescribeReportCreationInput$, () => DescribeReportCreationOutput$
];
var GetComplianceSummary$ = [9, n0, _GCS,
    { [_h]: ["POST", "/GetComplianceSummary", 200] }, () => GetComplianceSummaryInput$, () => GetComplianceSummaryOutput$
];
var GetResources$ = [9, n0, _GR,
    { [_h]: ["POST", "/GetResources", 200] }, () => GetResourcesInput$, () => GetResourcesOutput$
];
var GetTagKeys$ = [9, n0, _GTK,
    { [_h]: ["POST", "/GetTagKeys", 200] }, () => GetTagKeysInput$, () => GetTagKeysOutput$
];
var GetTagValues$ = [9, n0, _GTV,
    { [_h]: ["POST", "/GetTagValues", 200] }, () => GetTagValuesInput$, () => GetTagValuesOutput$
];
var ListRequiredTags$ = [9, n0, _LRT,
    { [_h]: ["POST", "/ListRequiredTags", 200] }, () => ListRequiredTagsInput$, () => ListRequiredTagsOutput$
];
var StartReportCreation$ = [9, n0, _SRC,
    { [_h]: ["POST", "/StartReportCreation", 200] }, () => StartReportCreationInput$, () => StartReportCreationOutput$
];
var TagResources$ = [9, n0, _TR,
    { [_h]: ["POST", "/TagResources", 200] }, () => TagResourcesInput$, () => TagResourcesOutput$
];
var UntagResources$ = [9, n0, _UR,
    { [_h]: ["POST", "/UntagResources", 200] }, () => UntagResourcesInput$, () => UntagResourcesOutput$
];

class DescribeReportCreationCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("ResourceGroupsTaggingAPI_20170126", "DescribeReportCreation", {})
    .n("ResourceGroupsTaggingAPIClient", "DescribeReportCreationCommand")
    .sc(DescribeReportCreation$)
    .build() {
}

class GetComplianceSummaryCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("ResourceGroupsTaggingAPI_20170126", "GetComplianceSummary", {})
    .n("ResourceGroupsTaggingAPIClient", "GetComplianceSummaryCommand")
    .sc(GetComplianceSummary$)
    .build() {
}

class GetResourcesCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("ResourceGroupsTaggingAPI_20170126", "GetResources", {})
    .n("ResourceGroupsTaggingAPIClient", "GetResourcesCommand")
    .sc(GetResources$)
    .build() {
}

class GetTagKeysCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("ResourceGroupsTaggingAPI_20170126", "GetTagKeys", {})
    .n("ResourceGroupsTaggingAPIClient", "GetTagKeysCommand")
    .sc(GetTagKeys$)
    .build() {
}

class GetTagValuesCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("ResourceGroupsTaggingAPI_20170126", "GetTagValues", {})
    .n("ResourceGroupsTaggingAPIClient", "GetTagValuesCommand")
    .sc(GetTagValues$)
    .build() {
}

class ListRequiredTagsCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("ResourceGroupsTaggingAPI_20170126", "ListRequiredTags", {})
    .n("ResourceGroupsTaggingAPIClient", "ListRequiredTagsCommand")
    .sc(ListRequiredTags$)
    .build() {
}

class StartReportCreationCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("ResourceGroupsTaggingAPI_20170126", "StartReportCreation", {})
    .n("ResourceGroupsTaggingAPIClient", "StartReportCreationCommand")
    .sc(StartReportCreation$)
    .build() {
}

class TagResourcesCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("ResourceGroupsTaggingAPI_20170126", "TagResources", {})
    .n("ResourceGroupsTaggingAPIClient", "TagResourcesCommand")
    .sc(TagResources$)
    .build() {
}

class UntagResourcesCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("ResourceGroupsTaggingAPI_20170126", "UntagResources", {})
    .n("ResourceGroupsTaggingAPIClient", "UntagResourcesCommand")
    .sc(UntagResources$)
    .build() {
}

const commands = {
    DescribeReportCreationCommand,
    GetComplianceSummaryCommand,
    GetResourcesCommand,
    GetTagKeysCommand,
    GetTagValuesCommand,
    ListRequiredTagsCommand,
    StartReportCreationCommand,
    TagResourcesCommand,
    UntagResourcesCommand,
};
class ResourceGroupsTaggingAPI extends ResourceGroupsTaggingAPIClient {
}
smithyClient.createAggregatedClient(commands, ResourceGroupsTaggingAPI);

const paginateGetComplianceSummary = core.createPaginator(ResourceGroupsTaggingAPIClient, GetComplianceSummaryCommand, "PaginationToken", "PaginationToken", "MaxResults");

const paginateGetResources = core.createPaginator(ResourceGroupsTaggingAPIClient, GetResourcesCommand, "PaginationToken", "PaginationToken", "ResourcesPerPage");

const paginateGetTagKeys = core.createPaginator(ResourceGroupsTaggingAPIClient, GetTagKeysCommand, "PaginationToken", "PaginationToken", "");

const paginateGetTagValues = core.createPaginator(ResourceGroupsTaggingAPIClient, GetTagValuesCommand, "PaginationToken", "PaginationToken", "");

const paginateListRequiredTags = core.createPaginator(ResourceGroupsTaggingAPIClient, ListRequiredTagsCommand, "NextToken", "NextToken", "MaxResults");

const ErrorCode = {
    INTERNAL_SERVICE_EXCEPTION: "InternalServiceException",
    INVALID_PARAMETER_EXCEPTION: "InvalidParameterException",
};
const GroupByAttribute = {
    REGION: "REGION",
    RESOURCE_TYPE: "RESOURCE_TYPE",
    TARGET_ID: "TARGET_ID",
};
const TargetIdType = {
    ACCOUNT: "ACCOUNT",
    OU: "OU",
    ROOT: "ROOT",
};

Object.defineProperty(exports, "$Command", {
    enumerable: true,
    get: function () { return smithyClient.Command; }
});
Object.defineProperty(exports, "__Client", {
    enumerable: true,
    get: function () { return smithyClient.Client; }
});
exports.ComplianceDetails$ = ComplianceDetails$;
exports.ConcurrentModificationException = ConcurrentModificationException;
exports.ConcurrentModificationException$ = ConcurrentModificationException$;
exports.ConstraintViolationException = ConstraintViolationException;
exports.ConstraintViolationException$ = ConstraintViolationException$;
exports.DescribeReportCreation$ = DescribeReportCreation$;
exports.DescribeReportCreationCommand = DescribeReportCreationCommand;
exports.DescribeReportCreationInput$ = DescribeReportCreationInput$;
exports.DescribeReportCreationOutput$ = DescribeReportCreationOutput$;
exports.ErrorCode = ErrorCode;
exports.FailureInfo$ = FailureInfo$;
exports.GetComplianceSummary$ = GetComplianceSummary$;
exports.GetComplianceSummaryCommand = GetComplianceSummaryCommand;
exports.GetComplianceSummaryInput$ = GetComplianceSummaryInput$;
exports.GetComplianceSummaryOutput$ = GetComplianceSummaryOutput$;
exports.GetResources$ = GetResources$;
exports.GetResourcesCommand = GetResourcesCommand;
exports.GetResourcesInput$ = GetResourcesInput$;
exports.GetResourcesOutput$ = GetResourcesOutput$;
exports.GetTagKeys$ = GetTagKeys$;
exports.GetTagKeysCommand = GetTagKeysCommand;
exports.GetTagKeysInput$ = GetTagKeysInput$;
exports.GetTagKeysOutput$ = GetTagKeysOutput$;
exports.GetTagValues$ = GetTagValues$;
exports.GetTagValuesCommand = GetTagValuesCommand;
exports.GetTagValuesInput$ = GetTagValuesInput$;
exports.GetTagValuesOutput$ = GetTagValuesOutput$;
exports.GroupByAttribute = GroupByAttribute;
exports.InternalServiceException = InternalServiceException;
exports.InternalServiceException$ = InternalServiceException$;
exports.InvalidParameterException = InvalidParameterException;
exports.InvalidParameterException$ = InvalidParameterException$;
exports.ListRequiredTags$ = ListRequiredTags$;
exports.ListRequiredTagsCommand = ListRequiredTagsCommand;
exports.ListRequiredTagsInput$ = ListRequiredTagsInput$;
exports.ListRequiredTagsOutput$ = ListRequiredTagsOutput$;
exports.PaginationTokenExpiredException = PaginationTokenExpiredException;
exports.PaginationTokenExpiredException$ = PaginationTokenExpiredException$;
exports.RequiredTag$ = RequiredTag$;
exports.ResourceGroupsTaggingAPI = ResourceGroupsTaggingAPI;
exports.ResourceGroupsTaggingAPIClient = ResourceGroupsTaggingAPIClient;
exports.ResourceGroupsTaggingAPIServiceException = ResourceGroupsTaggingAPIServiceException;
exports.ResourceGroupsTaggingAPIServiceException$ = ResourceGroupsTaggingAPIServiceException$;
exports.ResourceTagMapping$ = ResourceTagMapping$;
exports.StartReportCreation$ = StartReportCreation$;
exports.StartReportCreationCommand = StartReportCreationCommand;
exports.StartReportCreationInput$ = StartReportCreationInput$;
exports.StartReportCreationOutput$ = StartReportCreationOutput$;
exports.Summary$ = Summary$;
exports.Tag$ = Tag$;
exports.TagFilter$ = TagFilter$;
exports.TagResources$ = TagResources$;
exports.TagResourcesCommand = TagResourcesCommand;
exports.TagResourcesInput$ = TagResourcesInput$;
exports.TagResourcesOutput$ = TagResourcesOutput$;
exports.TargetIdType = TargetIdType;
exports.ThrottledException = ThrottledException;
exports.ThrottledException$ = ThrottledException$;
exports.UntagResources$ = UntagResources$;
exports.UntagResourcesCommand = UntagResourcesCommand;
exports.UntagResourcesInput$ = UntagResourcesInput$;
exports.UntagResourcesOutput$ = UntagResourcesOutput$;
exports.paginateGetComplianceSummary = paginateGetComplianceSummary;
exports.paginateGetResources = paginateGetResources;
exports.paginateGetTagKeys = paginateGetTagKeys;
exports.paginateGetTagValues = paginateGetTagValues;
exports.paginateListRequiredTags = paginateListRequiredTags;
